pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/PhysicianInstitutionComponent.java

```diff
@@ -10,66 +18,106 @@
-                                            boolean showFields,
-                                            boolean required,
-                                            boolean hideNumber) {
+    protected PhysicianInstitutionComponent(
+            ComponentType physicianOrInstitution,
+            InstitutionPhysicianComponentDto instDto,
+            boolean shouldHideNumber
+    ) {
         super(physicianOrInstitution);
         if (physicianOrInstitution != ComponentType.PHYSICIAN && physicianOrInstitution != ComponentType.INSTITUTION) {
             throw new IllegalArgumentException("Physician/Institution component must be either " + ComponentType
                     .PHYSICIAN + " or " + ComponentType.INSTITUTION);
         }
-        this.allowMultiple = allowMultiple;
-        this.addButtonText = addButtonText;
-        this.titleText = titleText;
-        this.subtitleText = subtitleText;
-        this.institutionType = institutionType;
-        this.showFields = showFields;
-        this.hideDisplayNumber = hideNumber;
-        this.required = required;
-        initParametersMap();
+        this.instDto = instDto;
+        this.hideDisplayNumber = shouldHideNumber;
     }
 
-    private final void initParametersMap() {
-        parameters.put(ALLOW_MULTIPLE, allowMultiple);
-        parameters.put(ADD_BUTTON_TEXT, addButtonText);
-        parameters.put(TITLE_TEXT, titleText);
-        parameters.put(SUBTITLE_TEXT, subtitleText);
-        parameters.put(INSTITUTION_TYPE, institutionType.name());
-        parameters.put(SHOW_FIELDS, showFields);
-        parameters.put(REQUIRED, required);
+    private final void setSerializedFields() {
+        serializedFields.setAllowMultiple(instDto.getAllowMultiple());
+        serializedFields.setButtonText(buttonText);
+        serializedFields.setTitleText(titleText);
+        serializedFields.setSubtitleText(subtitleText);
+        serializedFields.setInstitutionType(instDto.getInstitutionType().name());
+        serializedFields.setShowFields(instDto.showFields());
+        serializedFields.setIsRequired(instDto.isRequired());
+    }
+
+    @Override
+    public void registerTemplateIds(Consumer<Long> registry) {
+        super.registerTemplateIds(registry);
+        Optional.ofNullable(instDto.getButtonTemplateId()).ifPresentOrElse(registry::accept, () -> { });
+        Optional.ofNullable(instDto.getTitleTemplateId()).ifPresentOrElse(registry::accept, () -> { });
+        Optional.ofNullable(instDto.getSubtitleTemplateId()).ifPresentOrElse(registry::accept, () -> { });
+    }
+
+    @Override
+    public void applyRenderedTemplates(Provider<String> rendered, ContentStyle style) {
+        super.applyRenderedTemplates(rendered, style);
+        buttonText = rendered.get(instDto.getButtonTemplateId());
+        titleText = rendered.get(instDto.getTitleTemplateId());
+        subtitleText = rendered.get(instDto.getSubtitleTemplateId());
+
+        setSerializedFields();
```

Comment:
Since we are going to externally render the templates, this seems to be a good point to call `setSerializedFields()`. Only then the object can be considered to be "ready".